File: app/src/main/java/navumchyk/aliaksandr/cleanarchitecture/ui/adapters/common/BaseAdapter.java
package navumchyk.aliaksandr.cleanarchitecture.ui.adapters.common;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;

import rx.subjects.PublishSubject;

/**
 * @author Aliaksandr Navumchyk
 */
public abstract class BaseAdapter<TypeData, TypeHolder extends RecyclerView.ViewHolder, ClickData>
        extends RecyclerView.Adapter<TypeHolder> implements IBaseAdapter<TypeData, ClickData> {

    private PublishSubject<ClickData> mClick = PublishSubject.create();
    private ArrayList<TypeData> mData = new ArrayList<>();
    private int mLayoutResId;

    public BaseAdapter(int layoutResId) {
        mLayoutResId = layoutResId;
    }

    public abstract TypeHolder createViewHolder(View view,
                                                ViewGroup viewGroup,
                                                PublishSubject<ClickData> click);

    @Override
    public TypeHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
        final View view = LayoutInflater
                .from(viewGroup.getContext()).inflate(mLayoutResId, viewGroup, false);
        return createViewHolder(view, viewGroup, mClick);
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    @Override
    public void onViewDetachedFromWindow(TypeHolder holder) {
        final BaseHolder baseHolder = (BaseHolder) holder;
        baseHolder.unbindView();
        super.onViewDetachedFromWindow(holder);
    }

    @Override
    public void setData(@NonNull ArrayList<TypeData> data) {
        mData = data;
        this.notifyDataSetChanged();
    }

    @Override
    public ArrayList<TypeData> getData() {
        return mData;
    }

    @Override
    public TypeData getItem(int position) {
        return mData.get(position);
    }

    @Override
    public PublishSubject<ClickData> observeItemClick() {
        return mClick;
    }
}

File: app/src/main/java/navumchyk/aliaksandr/cleanarchitecture/ui/adapters/common/BaseAdapter.kt
package navumchyk.aliaksandr.cleanarchitecture.ui.adapters.common

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import rx.subjects.PublishSubject
import java.util.*

/**
 * @author Aliaksandr Navumchyk
 */
abstract class BaseAdapter<TypeData, TypeHolder : RecyclerView.ViewHolder, ClickData>(
        private val mLayoutResId: Int) : RecyclerView.Adapter<TypeHolder>(),
        IBaseAdapter<TypeData, ClickData> {

    private val mClick = PublishSubject.create<ClickData>()
    private var mData: ArrayList<TypeData> = ArrayList()

    abstract fun createViewHolder(view: View,
                                  viewGroup: ViewGroup,
                                  click: PublishSubject<ClickData>): TypeHolder

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): TypeHolder {
        val view = LayoutInflater
                .from(viewGroup.context).inflate(mLayoutResId, viewGroup, false)
        return createViewHolder(view, viewGroup, mClick)
    }

    override fun getItemCount(): Int {
        return mData.size
    }

    override fun onViewDetachedFromWindow(holder: TypeHolder) {
        val baseHolder = holder as BaseHolder<*>
        baseHolder.unbindView()
        super.onViewDetachedFromWindow(holder)
    }

    override fun setData(data: ArrayList<TypeData>) {
        mData = data
        this.notifyDataSetChanged()
    }

    override fun getData(): ArrayList<TypeData> {
        return mData
    }

    override fun getItem(position: Int): TypeData {
        return mData[position]
    }

    override fun observeItemClick(): PublishSubject<ClickData> {
        return mClick
    }
}